File: client/src/main/java/com/flipkart/grayskull/auth/BasicAuthHeaderProvider.java
package com.flipkart.grayskull.auth;

import java.nio.charset.StandardCharsets;
import java.util.Base64;

/**
 * Basic authentication header provider implementation.
 * <p>
 * This implementation provides HTTP Basic Authentication by encoding the username
 * and password in Base64 format according to RFC 7617.
 * </p>
 */
public final class BasicAuthHeaderProvider implements GrayskullAuthHeaderProvider {
    
    private final String username;
    private final String password;
    
    public BasicAuthHeaderProvider(String username, String password) {
        if (username == null || password == null) {
            throw new IllegalArgumentException("Username and password cannot be null");
        }
        this.username = username;
        this.password = password;
    }

    @Override
    public String getAuthHeader() {
        String credentials = username + ":" + password;
        String encodedCredentials = Base64.getEncoder()
            .encodeToString(credentials.getBytes(StandardCharsets.UTF_8));
        return "Basic " + encodedCredentials;
    }
}

File: clients/go/client-impl/auth/basicAuthHeaderProvider.go
package auth

import (
	"encoding/base64"
	"fmt"
	"strings"
)

// BasicAuthHeaderProvider implements GrayskullAuthHeaderProvider for HTTP Basic Authentication.
//
// This implementation provides HTTP Basic Authentication by encoding the username
// and password in Base64 format according to RFC 7617.
type BasicAuthHeaderProvider struct {
	username string
	password string
}

// NewBasicAuthHeaderProvider creates a new instance of BasicAuthHeaderProvider.
// Returns an error if username or password is empty.
func NewBasicAuthHeaderProvider(username, password string) (*BasicAuthHeaderProvider, error) {
	username = strings.TrimSpace(username)
	if username == "" || password == "" {
		return nil, fmt.Errorf("username and password cannot be empty")
	}
	return &BasicAuthHeaderProvider{
		username: username,
		password: password,
	}, nil
}

// GetAuthHeader returns the Basic Authentication header value.
func (b *BasicAuthHeaderProvider) GetAuthHeader() (string, error) {
	credentials := fmt.Sprintf("%s:%s", b.username, b.password)
	encoded := base64.StdEncoding.EncodeToString([]byte(credentials))
	return "Basic " + encoded, nil
}
